File: java/127.word-ladder.java
/*
 * @lc app=leetcode id=127 lang=java
 *
 * [127] Word Ladder
 */
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        if (wordList == null || wordList.size() == 0) return 0;
        int d = 2;
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        boolean[] visited = new boolean[wordList.size()];
        while (!queue.isEmpty()) {
            int len = queue.size();
            for (int i = 0; i < len; ++i) {
                String word = queue.poll();
                for (int idx = 0; idx < wordList.size(); ++idx) {
                    if (visited[idx]) continue;
                    String cand = wordList.get(idx);
                    if (dist(word, cand) == 1) {
                        if (cand.equals(endWord)) {
                            return d;
                        }
                        visited[idx] = true;
                        queue.offer(cand);
                    }
                }
            }
            d++;
        }
        return 0;
    }

    private int dist(String w1, String w2) {
        int d = 0;
        if (w1.equals(w2)) return 0;
        for (int i  = 0; i < w1.length(); ++i) {
            if (w1.charAt(i) != w2.charAt(i)) d++;
        }
        return d;
    }
}

File: go/127.word-ladder.go
/*
 * @lc app=leetcode id=127 lang=golang
 *
 * [127] Word Ladder
 */

// @lc code=start
func ladderLength(beginWord string, endWord string, wordList []string) int {
	dict := initDict(wordList)
	queue := []int{-1}
	step, curcnt, nextcnt := 1, 1, 0
	for len(queue) != 0 {
		idx := queue[0]
		queue = queue[1:]

		word := beginWord
		if idx != -1 {
			word = wordList[idx]
		}
		if word == endWord {
			return step
		}

		indexes := idxToAppend(word, dict)
		queue = append(queue, indexes...)
		nextcnt += len(indexes)

		if curcnt == 1 {
			step, curcnt, nextcnt = step+1, nextcnt, 0
		} else {
			curcnt -= 1
		}
	}
	return 0
}

func idxToAppend(word string, dict map[string]int) []int {
	res := make([]int, 0)
	bytes := []byte(word)
	for i := 0; i < len(bytes); i++ {
		ob := bytes[i]
		for c := 0; c < 26; c++ {
			if nb := byte('a'+c); nb != ob {
				bytes[i] = nb
				str := string(bytes)
				if idx, b := dict[str]; b {
					res = append(res, idx)
					delete(dict, str)
				}
			}
		}
		bytes[i] = ob
	}
	return res
}

func initDict(wordList []string) map[string]int {
	dict := make(map[string]int)
	for i, s := range wordList {
		dict[s] = i
	}
	return dict
}
// @lc code=end

